runner/logs.go

```diff
@@ -111,73 +313,126 @@ func TestLogsForAllServers(servers []*state.Server, globalCollectionOpts state.C
-			if err != nil {
-				hasFailedServers = true
-				prefixedLogger.PrintError("ERROR - Could not get Amazon RDS logs: %s", err)
+				success = true
 			} else {
-				prefixedLogger.PrintInfo("  Log test successful")
+				success = false
 			}
+		} else if server.Config.AwsDbInstanceID != "" {
+			success = testRdsLogDownload(ctx, &wg, server, globalCollectionOpts, prefixedLogger)
 		} else if server.Config.AzureDbServerName != "" && server.Config.AzureEventhubNamespace != "" && server.Config.AzureEventhubName != "" {
-			prefixedLogger.PrintInfo("Testing log collection (Azure Database)...")
-			err := azure.LogTestRun(server, globalCollectionOpts, prefixedLogger)
-			if err != nil {
-				hasFailedServers = true
-				prefixedLogger.PrintError("ERROR - Could not get logs through Azure Event Hub: %s", err)
-			} else {
-				prefixedLogger.PrintInfo("  Log test successful")
-			}
+			success = testAzureLogStream(ctx, &wg, server, globalCollectionOpts, prefixedLogger)
 		} else if server.Config.GcpCloudSQLInstanceID != "" && server.Config.GcpPubsubSubscription != "" {
-			prefixedLogger.PrintInfo("Testing log collection (Google Cloud SQL)...")
-			err := google_cloudsql.LogTestRun(server, globalCollectionOpts, prefixedLogger)
-			if err != nil {
-				hasFailedServers = true
-				prefixedLogger.PrintError("ERROR - Could not get logs through Google Cloud Pub/Sub: %s", err)
-			} else {
-				prefixedLogger.PrintInfo("  Log test successful")
-			}
+			success = testGoogleCloudsqlLogStream(ctx, &wg, server, globalCollectionOpts, prefixedLogger)
+		}
+
+		if !success {
+			hasFailedServers = true
 		}
+
+		cancel()
 	}
 
 	return
 }
 
-// DownloadLogsFromAllServers - Downloads logs from all servers that are remote systems and sends them to the pganalyze service
-func DownloadLogsFromAllServers(servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
-	var wg sync.WaitGroup
+func testLocalLogTail(ctx context.Context, wg *sync.WaitGroup, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) bool {
+	logger.PrintInfo("Testing log collection (local)...")
 
-	if !globalCollectionOpts.CollectLogs {
-		return
+	logTestSucceeded := make(chan bool, 1)
+	parsedLogStream := setupLogStreamer(ctx, wg, globalCollectionOpts, logger, []*state.Server{server}, logTestSucceeded, stream.LogTestCollectorIdentify)
+
+	err := selfhosted.SetupLogTailForServer(ctx, wg, globalCollectionOpts, logger, server, parsedLogStream)
+	if err != nil {
+		logger.PrintError("ERROR - Could not tail logs for server: %s", err)
+		return false
 	}
 
-	for _, server := range servers {
-		if server.Config.DisableLogs || (server.Grant.Valid && !server.Grant.Config.EnableLogs) {
-			continue
-		}
+	logs.EmitTestLogMsg(server, globalCollectionOpts, logger)
 
-		if server.Config.AwsDbInstanceID == "" {
-			continue
-		}
+	select {
+	case <-logTestSucceeded:
+		break // TODO: Is this correct?
```

Comment:
?

Reply:
Good catch, that was a left over - I was trying to reason whether its safe to have `select` here without a `for` loop, but the answer is yes, this code is fine as is (see e.g. https://stackoverflow.com/questions/39212333/how-can-i-use-time-after-and-default-in-golang)